lib/utils/ensure-single-browser-tab-only.js

```diff
@@ -0,0 +1,69 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import BootWarning from '../components/boot-warning';
+
+const HEARTBEAT_DELAY = 1000;
+const clientId = uuidv4();
+const emptyLock = [null, -Infinity];
+const foundElectron = window.process && window.process.type;
+
+if (!foundElectron) {
+  if ('lock-acquired' !== grabSessionLock()) {
+    ReactDOM.render(<BootWarning />, document.getElementById('root'));
+    throw new Error('Simplenote can only be opened in one tab');
+  }
+
+  loop(() => {
+    switch (grabSessionLock()) {
+      case 'lock-acquired':
+        return true; // keep updating the lock and look for other sessions which may have taken it
+
+      default:
+        return false; // stop watching - the user can proceed at their own risk
+    }
+  });
+
+  window.addEventListener('unload', function(event) {
+    localStorage.removeItem('session-lock');
+  });
```

Comment:
we need to be careful here. it's a good point you bring up.

while this isn't needed to free the lock generally, you've pointed out that we have the opportunity to free it for ourselves in some or most cases

I'm worried that in unloading we could run after another session has grabbed the lock and then destroy that session. we may not be able to eliminate this but we can probably add one tiny guard to mitigate most races

```js
let keepGoing = true;

…
loop(() => {
	if ( ! keepGoing ) {
		return false;
	}

	…
} );
…

window.addEventListener('unload', function(event) {
	keepGoing = false;
	const [ lastClient ] = localStorage.getItem( 'session-lock' );
	lastClient === clientId && localStorage.removeItem( 'session-lock' );
} );
```

this is based on the untested assumption that we might incur a delay between killing the page, shutting down our timers, and running the `unload` code

probably also responsible to make sure we don't finish running our event queue and inadvertently grab the lock after we've killed the page.